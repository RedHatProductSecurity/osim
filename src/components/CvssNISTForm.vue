<script setup lang="ts">
import { computed } from 'vue';
import Modal from '@/components/widgets/Modal.vue';

import LabelTextarea from '@/components/widgets/LabelTextarea.vue';
import { useUserStore } from '@/stores/UserStore';
import { useModal } from '@/composables/useModal';

const props = defineProps<{
  cveid?: string | null;
  summary?: string | null;
  bugzilla?: string;
  cvss?: string;
  nistcvss?: string;
}>();

const userStore = useUserStore();
const { isModalOpen, openModal, closeModal } = useModal();

const toEmail = 'nvd@nist.gov';
const ccEmail = 'secalert@redhat.com';
const subject = computed(() => `CVSS Rescore Request - ${props.cveid}`);
const emailBody = computed(() => `Hello,

I have performed an analysis of ${props.cveid} on behalf of Red Hat Product Security,
resulting in a Red Hat CVSS score which is different from the NIST score.
Our information and analysis is included below,
and we would appreciate your consideration and review.

CVE : ${props.cveid}

Red Hat Bugzilla: ${props.bugzilla}
NVD Page: https://nvd.nist.gov/vuln/detail/${props.cveid}
Red Hat CVSS: ${props.cvss}
NIST CVSS: ${props.nistcvss}

Flaw Summary: 
${props.summary}

Red Hat's CVSS Justification:
____

Thank you,
-${userStore.userEmail}
`);

function openMailto() {
  const recipient = encodeURI(toEmail);
  const cc = encodeURI(ccEmail);
  const encodedSubject = encodeURIComponent(subject.value);
  const encodedBody = encodeURIComponent(emailBody.value);
  const mailto = `mailto:${recipient}?cc=${cc}&subject=${encodedSubject}&body=${encodedBody}`;
  window.open(mailto, '_blank');
}
</script>

<template>
  <div class="osim-cve-nist-button">
    <button type="button" class="btn btn-secondary" @click="openModal">
      Email NIST
    </button>

    <Modal :show="isModalOpen" @close="closeModal">
      <template #title>
        Email NIST
      </template>
      <template #body>
        <div class="osim-input mb-3 border-start ps-3">
          <p><span class="fw-bold">To:</span> <span class="to-email">{{ toEmail }}</span></p>
          <p class="from-email"><span class="fw-bold">From:</span> {{ userStore.userEmail }}</p>
          <p><span class="fw-bold">CC:</span> <span class="cc-email">{{ ccEmail }}</span></p>
          <hr />
          <p>Subject:</p>
          <div class="mb-2">
            <input
              class="form-control"
              :readonly="true"
              type="text"
              :value="subject"
            />
          </div>
          <hr />
        </div>

        <LabelTextarea v-model="emailBody" label="Body:" />
      </template>
      <template #footer>
        <button
          type="button"
          class="btn btn-secondary cancel-btn"
          data-bs-dismiss="modal"
          @click="closeModal"
        >Cancel
        </button>
        <button type="button" class="btn btn-primary send-email" @click.prevent="openMailto">
          Send Email
        </button>
      </template>
    </Modal>
  </div>
</template>

<style scoped>
.osim-cve-nist-button :deep(.modal-dialog) {
  max-width: 80ch;
}
</style>
