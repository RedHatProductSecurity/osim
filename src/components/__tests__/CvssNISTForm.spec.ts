import { mount, flushPromises } from '@vue/test-utils';
import { describe, it, expect } from 'vitest';

import CvssNISTForm from '../CvssNISTForm.vue';

import { createTestingPinia } from '@pinia/testing';

describe('CvssNISTForm', () => {
  it('renders a button', () => {
    const pinia = createTestingPinia({
      createSpy: vitest.fn,
      stubActions: false,
    });
    const wrapper = mount(CvssNISTForm, {
      props: {
        flaw: 'any',
        cveid: 'string',
        flawSummary: 'string',
        bugzilla: 'string',
        nvdpage: 'string',
        cvss: 'string',
        nistcvss: 'string',
        cvssjustification: 'string',
      },
      global: {
        plugins: [
          pinia
        ]
      }
    });
    const button = wrapper.find('button');
    expect(button.exists()).toBeTruthy();
    expect(button.text()).toBe('Email NIST');
  });

  it('does not render modal before clicking the button', () => {
    const pinia = createTestingPinia({
      createSpy: vitest.fn,
      stubActions: false,
    });
    const wrapper = mount(CvssNISTForm, {
      props: {
        flaw: 'any',
        cveid: 'string',
        flawSummary: 'string',
        bugzilla: 'string',
        nvdpage: 'string',
        cvss: 'string',
        nistcvss: 'string',
        cvssjustification: 'string',
      },
      global: {
        plugins: [
          pinia,
        ]
      }
    });
    const button = wrapper.find('button');
    expect(button.exists()).toBeTruthy();
    expect(wrapper.find('div.modal-content').exists()).toBe(false);
  });

  it('renders a modal when clicking the button', async () => {
    const pinia = createTestingPinia({
      createSpy: vitest.fn,
      stubActions: false,
    });
    const wrapper = mount(CvssNISTForm, {
      props: {
        flaw: 'any',
        cveid: 'string',
        flawSummary: 'string',
        bugzilla: 'string',
        nvdpage: 'string',
        cvss: 'string',
        nistcvss: 'string',
        cvssjustification: 'string',
      },
      global: {
        plugins: [
          pinia,
        ]
      }
    });
    const button = wrapper.find('button');
    await button.trigger('click');
    await flushPromises();

    expect(wrapper.find('div.modal-content').exists()).toBeTruthy();
    expect(wrapper.find('input.form-control').exists()).toBeTruthy();
    expect(wrapper.find('textarea.form-control').exists()).toBeTruthy();
    expect(wrapper.find('label.osim-input').exists()).toBeTruthy();
    expect(wrapper.find('button.cancel-btn').text()).toBe('Cancel');
    expect(wrapper.find('button.send-email').text()).toBe('Send Email');
    expect(wrapper.find('span.to-email').text()).toBe('nvd@nist.gov');
    expect(wrapper.find('span.cc-email').text()).toBe('secalert@redhat.com');
    expect(wrapper.find('p.from-email').exists()).toBeTruthy();
  });

  it('close modal by clicking cancel button', async () => {
    const pinia = createTestingPinia({
      createSpy: vitest.fn,
      stubActions: false,
    });
    const wrapper = mount(CvssNISTForm, {
      props: {
        flaw: 'any',
        cveid: 'string',
        flawSummary: 'string',
        bugzilla: 'string',
        nvdpage: 'string',
        cvss: 'string',
        nistcvss: 'string',
        cvssjustification: 'string',
      },
      global: {
        plugins: [
          pinia,
        ]
      }
    });
    const button = wrapper.find('button');
    await button.trigger('click');
    await flushPromises();

    expect(wrapper.find('div.modal-content').exists()).toBeTruthy();
    const closeButton = wrapper.find('button.cancel-btn');
    expect(closeButton.exists()).toBeTruthy();
    expect(closeButton.text()).toBe('Cancel');
    await closeButton.trigger('click');
    await flushPromises();
    expect(wrapper.find('div.modal-content').exists()).toBeFalsy();
  });

  it('open new window by clicking send button', async () => {
    const pinia = createTestingPinia({
      createSpy: vitest.fn,
      stubActions: false,
    });
    const wrapper = mount(CvssNISTForm, {
      props: {
        flaw: 'any',
        cveid: 'string',
        flawSummary: 'string',
        bugzilla: 'string',
        nvdpage: 'string',
        cvss: 'string',
        nistcvss: 'string',
        cvssjustification: 'string',
      },
      global: {
        plugins: [
          pinia,
        ]
      }
    });
    const button = wrapper.find('button');
    await button.trigger('click');
    await flushPromises();
    const spy = vi.spyOn(window, 'open');

    expect(wrapper.find('div.modal-content').exists()).toBeTruthy();
    const sendButton = wrapper.find('button.send-email');
    expect(sendButton.exists()).toBeTruthy();
    await sendButton.trigger('click');
    expect(spy).toHaveBeenCalledOnce();
  });
});
