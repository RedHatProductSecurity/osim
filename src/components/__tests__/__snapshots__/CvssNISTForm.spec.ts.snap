// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cvssNISTForm > renders a button 1`] = `
"<div data-v-b05c734f="" class="osim-cve-nist-form" flaw="any" nvdpage="string" cvssjustification="string"><button data-v-b05c734f="" type="button" class="osim-cve-email btn btn-secondary"> Email NIST </button>
  <transition-stub data-v-296c460d="" name="modal" appear="false" persisted="false" css="true">
    <!--v-if-->
  </transition-stub>
  <transition-stub data-v-296c460d="" name="modal-bg" appear="false" persisted="false" css="true">
    <!--v-if-->
  </transition-stub>
</div>"
`;

exports[`cvssNISTForm > renders a modal when clicking the button 1`] = `
"<div data-v-b05c734f="" class="osim-cve-nist-form" flaw="any" nvdpage="string" cvssjustification="string"><button data-v-b05c734f="" type="button" class="osim-cve-email btn btn-secondary"> Email NIST </button>
  <transition-stub data-v-296c460d="" name="modal" appear="false" persisted="false" css="true">
    <div data-v-296c460d="" class="modal fade show" style="display: block;" aria-hidden="false" tabindex="-1" aria-labelledby="modalTitle" role="dialog">
      <div data-v-296c460d="" class="modal-dialog">
        <div data-v-296c460d="" class="modal-content">
          <div data-v-296c460d="" class="modal-header">
            <h1 data-v-296c460d="" class="modal-title fs-5"> Email NIST </h1><button data-v-296c460d="" type="button" class="btn-close" aria-label="Close"></button>
          </div>
          <div data-v-296c460d="" class="modal-body">
            <div data-v-b05c734f="" class="osim-input mb-3 border-start ps-3">
              <p data-v-b05c734f=""><span data-v-b05c734f="" class="fw-bold">To:</span> <span data-v-b05c734f="" class="to-email">nvd@nist.gov</span></p>
              <p data-v-b05c734f="" class="from-email"><span data-v-b05c734f="" class="fw-bold">From:</span> </p>
              <p data-v-b05c734f=""><span data-v-b05c734f="" class="fw-bold">CC:</span> <span data-v-b05c734f="" class="cc-email">secalert@redhat.com</span></p>
              <hr data-v-b05c734f="">
              <p data-v-b05c734f="">Subject:</p>
              <div data-v-b05c734f="" class="mb-2"><input data-v-b05c734f="" class="form-control" readonly="" type="text" value="CVSS Rescore Request - string"></div>
              <hr data-v-b05c734f="">
            </div><label data-v-055401b4="" data-v-b05c734f="" class="osim-input mb-2">
              <div data-v-055401b4="" class="row px-0 mx-0"><span data-v-055401b4="" class="form-label col-3 position-relative"><!--v-if--> Body: <!--attrs: {{ $attrs }}--></span><!--    @focus="resizeTextarea"-->
                <!--    @keyup="resizeTextarea"--><textarea data-v-055401b4="" class="form-control col-9 d-inline-block" rows="5" value="Hello,

I have performed an analysis of string on behalf of Red Hat Product Security,
resulting in a Red Hat CVSS score which is different from the NIST score.
Our information and analysis is included below,
and we would appreciate your consideration and review.

CVE : string

Red Hat Bugzilla: string
NVD Page: https://nvd.nist.gov/vuln/detail/string
Red Hat CVSS: string
NIST CVSS: string

Flaw Summary: 
string

Red Hat's CVSS Justification:
____

Thank you,
-
"></textarea>
                <!--v-if-->
              </div>
            </label>
          </div>
          <div data-v-296c460d="" class="modal-footer"><button data-v-b05c734f="" type="button" class="btn btn-secondary cancel-btn" data-bs-dismiss="modal">Cancel </button><button data-v-b05c734f="" type="button" class="btn btn-primary send-email"> Send Email </button></div>
        </div>
      </div>
    </div>
  </transition-stub>
  <transition-stub data-v-296c460d="" name="modal-bg" appear="false" persisted="false" css="true">
    <div data-v-296c460d="" class="modal-backdrop fade show"></div>
  </transition-stub>
</div>"
`;
